rest/src/test/java/controller/CarControllerTest.java

```diff
@@ -0,0 +1,170 @@
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+
+import static org.mockito.Mockito.*;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = RestTestConfig.class)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+class CarControllerTest {
+
+    private static final String CARS_LIST_STORAGE_JSON = "car-list-storage.json";
+
+    @Autowired
+    private CarService carService;
+    @Autowired
+    private CarController carController;
+    @Autowired
+    private ObjectMapper objectMapper;
+    @Autowired
+    private RestExceptionHandler restExceptionHandler;
+    private List<Car> carList;
+    private MockMvc mockMvc;
+
+    @BeforeEach
+    void setup() throws IOException {
+        mockMvc = MockMvcBuilders.standaloneSetup(carController)
+                .setControllerAdvice(restExceptionHandler)
+                .build();
+        File file = new ClassPathResource(CARS_LIST_STORAGE_JSON).getFile();
+        carList = objectMapper.readValue(file, new TypeReference<List<Car>>() {
+        });
+    }
+
+    @Test
+    void getCars() throws Exception {
+        when(carService.getCars()).thenReturn(carList);
+        mockMvc.perform(get("/carSale/car"))
+                .andExpect(status().isOk())
+                .andExpect(content().contentType(MediaType.APPLICATION_JSON_UTF8_VALUE))
+                .andExpect(content().json(objectMapper.writeValueAsString(carList)));
+        verify(carService, times(1)).getCars();
+    }
+
+    @Test
+    void getCar() throws Exception {
+        int carId = 1;
+        Car car = carList.get(carId);
+        when(carService.getCar(carId)).thenReturn(car);
+        mockMvc.perform(get("/carSale/car/{carId}", carId))
+                .andExpect(status().isOk())
+                .andExpect(content().contentType(MediaType.APPLICATION_JSON_UTF8_VALUE))
+                .andExpect(content().json(objectMapper.writeValueAsString(car)));
+        verify(carService, times(1)).getCar(carId);
+    }
+
+    @Test
+    void getNotExistCarSale() throws Exception {
+        int carId = 21;
+        when(carService.getCar(carId)).thenThrow(NotFoundException.class);
+        mockMvc.perform(get("/carSale/car/{carId}", carId))
+                .andExpect(status().isNotFound());
+        verify(carService, times(1)).getCar(carId);
+    }
+
+    @Test
+    void addCar() throws Exception {
+        int carId = 1;
+        Car car = carList.get(carId);
+        when(carService.addCar(any(Car.class))).thenReturn(carId);
+        mockMvc.perform(post("/carSale/car")
```

Comment:
Propose to replace it as a constant

Reply:
Are you talking about endpoints or carId?

Reply:
Given comment related to the row with URL
So, I mentioned about part of endpoint URL